### PR TITLE
Add Disabling Room Intro Scroll Capability

### PR DESCRIPTION
**Pull request by Lord-Haji**
Extension of #183

### CODE REVIEW

**Comment by HoeenCoder** [comment]
@Lord-Haji did you get permission from @DeathlyPlays to do this?

**Comment by Lord-Haji** [comment]
I did this coz I didn't get any response

**Comment by HoeenCoder** [comment]
Add this as a /usetoken token too

File: spacialgaze-plugins/misc.js [hunk 1]
```diff
@@ -372,6 +372,13 @@ exports.commands = {
 			msg += '<button class="button" name="send" value="/emote add, ' + target[1] + ', ' + target[2] + '">Add emote</button></center>';
 			delete user.tokens[target[0]];
 			return SG.messageSeniorStaff(msg);
+		case 'disableintroscroll':
```
Comment (Lord-Haji) [inline spacialgaze-plugins/misc.js:375]: This is already there

Comment (HoeenCoder) [reply spacialgaze-plugins/misc.js:375]: oops

Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js]: Ok for this though, when the user tries to use the token check to make sure the room exists, and the room dosent already have its roomintro limit removed.

File: rooms.js [hunk 2]
```diff
@@ -1350,7 +1350,8 @@ class ChatRoom extends Room {
 	}
 	getIntroMessage(user) {
 		let message = '';
-		if (this.introMessage) message += '\n|raw|<div class="infobox infobox-roomintro"><div' + (!this.isOfficial ? ' class="infobox-limited"' : '') + '>' + this.introMessage.replace(/\n/g, '') + '</div>';
+		let scroll = Db.AllowScrolls.get(this.id, false);
+		if (this.introMessage) message += '\n|raw|<div class="infobox infobox-roomintro"><div' + (!scroll ? ' class="infobox-limited"' : '') + '>' + this.introMessage.replace(/\n/g, '') + '</div>';
```
Comment (HoeenCoder) [inline rooms.js]: Offical rooms should automatically not have scroll bars.

**Review by HoeenCoder — COMMENTED** [review]
Make it possible for admins to enable roomintroscroll for rooms too.

File: rooms.js [hunk 3]
```diff
@@ -1350,7 +1350,8 @@ class ChatRoom extends Room {
 	}
 	getIntroMessage(user) {
 		let message = '';
-		if (this.introMessage) message += '\n|raw|<div class="infobox infobox-roomintro"><div' + (!this.isOfficial ? ' class="infobox-limited"' : '') + '>' + this.introMessage.replace(/\n/g, '') + '</div>';
+		let scroll = Db.AllowScrolls.get(this.id, false);
+		if (this.introMessage) message += '\n|raw|<div class="infobox infobox-roomintro"><div' + ((this.isOfficial || !scroll) ? ' class="infobox-limited"' : '') + '>' + this.introMessage.replace(/\n/g, '') + '</div>';
```
Comment (HoeenCoder) [inline rooms.js]: You need to add a `!` before `this.isOfficial`.

File: spacialgaze-plugins/misc.js [hunk 4]
```diff
@@ -2,9 +2,9 @@
  * Miscellaneous commands
  *
  * Fixed/Improved upon by: The Run, HoeenHero, Mystifi and Lord Haji.
- *
  * Some of this code was borrowed from panpawn/jd/other contributors; as
  * such, credits go to them as well.
+ * @license MIT license
```
Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js:7]: Why are we adding a license tag? the whole repo is MIT already.

File: spacialgaze-plugins/misc.js (hunk 1 above)
Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js]: This still dosent check to see if the room exists, or if the room already has its scroll bar disabled.

File: spacialgaze-plugins/misc.js [hunk 5]
```diff
@@ -398,4 +405,21 @@ exports.commands = {
 		this.parse("/tour create " + target + ", roundrobin");
 	},
 	rtourhelp: ["/rtour [format] - Creates a round robin tournament in the format provided."],
+
+	disableintroscroll: function (target, room, user) {
```
Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js:411]: Error out if the scroll is already disabled.

File: spacialgaze-plugins/misc.js [hunk 6]
```diff
@@ -398,4 +405,21 @@ exports.commands = {
...
+		if (!this.can('lock')) return false;
+		if (!target) return this.errorReply("No Room Specified");
+		target = toId(target);
+		if (!Rooms(target)) return this.errorReply(`${target} is not a room`);
+		Db.AllowScrolls.set(target, true);
+	},
+	disableintroscrollhelp: ["/disableintroscroll [room] - Disables scroll bar preset in the room's roomintro."],
+	enableintroscroll: function (target, room, user) {
```
Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js]: Error out if the scroll is already enabled.

File: spacialgaze-plugins/misc.js (hunk 5 above)
Comment (HoeenCoder) [reply spacialgaze-plugins/misc.js:411]: Min permission should be `&`

File: spacialgaze-plugins/misc.js (hunk 6 above)
Comment (HoeenCoder) [reply spacialgaze-plugins/misc.js]: Min permission should be `&`

File: rooms.js (hunk 3 above)
Comment (Lord-Haji) [reply rooms.js]: Oops Typo :P

File: spacialgaze-plugins/misc.js (hunk 4 above)
Comment (Lord-Haji) [reply spacialgaze-plugins/misc.js:7]: Standardizations with other files ??

File: spacialgaze-plugins/misc.js (hunk 5 above)
Comment (Lord-Haji) [reply spacialgaze-plugins/misc.js:411]: its set to lock ?

File: spacialgaze-plugins/misc.js (hunk 4 above)
Comment (HoeenCoder) [reply spacialgaze-plugins/misc.js:7]: Ill check that out after I eat, you can wait on this atm then.

Edit: Your right, this is fine.

File: spacialgaze-plugins/misc.js (hunk 5 above)
Comment (HoeenCoder) [reply spacialgaze-plugins/misc.js:411]: Lock is `%`

**Comment by DeathlyPlays** [comment]
To clarify I didn't give him permission xD, but it's whatever....

**Comment by ghost** [comment]
at least its better

File: rooms.js [hunk 7]
```diff
@@ -1350,7 +1350,8 @@ class ChatRoom extends Room {
 	}
 	getIntroMessage(user) {
 		let message = '';
-		if (this.introMessage) message += '\n|raw|<div class="infobox infobox-roomintro"><div' + (!this.isOfficial ? ' class="infobox-limited"' : '') + '>' + this.introMessage.replace(/\n/g, '') + '</div>';
+		let scroll = Db.DisabledScrolls.get(this.id, false);
+		if (this.introMessage) message += '\n|raw|<div class="infobox infobox-roomintro"><div' + ((!this.isOfficial || !scroll) ? ' class="infobox-limited"' : '') + '>' + this.introMessage.replace(/\n/g, '') + '</div>';
```
Comment (HoeenCoder) [inline rooms.js]: This needs to be a ``&&`` check to work.

File: spacialgaze-plugins/misc.js [hunk 8]
```diff
@@ -372,6 +372,15 @@ exports.commands = {
 			msg += '<button class="button" name="send" value="/emote add, ' + target[1] + ', ' + target[2] + '">Add emote</button></center>';
 			delete user.tokens[target[0]];
 			return SG.messageSeniorStaff(msg);
+		case 'disableintroscroll':
+			if (!target[2]) return this.errorReply('/usetoken disableintroscroll, [room]');
+			let roomid = toId(target[1]);
+			if (Rooms[roomid]) return this.errorReply(`${roomid} is not a room.`);
```
Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js]: This check says:

`If the room exists error out.`

**Review by HoeenCoder — COMMENTED** [review]
You'll need to update eShop.js to handle the new token too. Which should just be adding it to the `toToken` function in that file, but check to make sure I didn't miss anything when I checked it at the time of this review.

File: rooms.js (hunk 7 above)
Comment (Lord-Haji) [reply rooms.js]: Why?

Comment (Lord-Haji) [reply rooms.js]: Scroll won't show up,if the room has scroll disabled or if it is official isn't it?

Comment (HoeenCoder) [reply rooms.js]: Yeah actually im wrong too, this should be `((this.isOfficial || scroll) ? '' : ' class="infobox-limited"')`

**Comment by DeathlyPlays** [comment]
Meh

**Comment by Lord-Haji** [comment]
Yeah, actually I am confusing myself too much coz exams sorry :P

**Comment by DeathlyPlays** [comment]
@Lord-Haji making statuses like this generally will have portions coded in the battle-engine.js and/or statuses.js.  Hence why this move is not recommended.

**Comment by HoeenCoder** [comment]
@DeathlyPlays what are you talking about? This PR has nothing to do with statues.js or battle-engine.js

**Comment by ghost** [comment]
He may have confused a different PR with this one @HoeenCoder

**Comment by Mystifi** [comment]
In my opinion, database names shouldn't be capitalized at all, similarly to how standard Config fields aren't. However, I'd much rather you, at least, use lowerCamelCase when naming databases, @Lord-Haji - if you could edit the code to use this, it would be appreciated.

File: spacialgaze-plugins/misc.js (hunk 8 above)
Comment (Mystifi) [reply spacialgaze-plugins/misc.js]: You are also trying to access `Rooms` as if it were an object; rather, it's a function. It should be `if (!Rooms(roomid))`, like it is later in your code.

Comment (Lord-Haji) [reply spacialgaze-plugins/misc.js]: It was typo, which was made by mistake as I was doing to too fast

**Comment by Lord-Haji** [comment]
tbh @Mystifi I prefer UpperCamelCase than lowerCamelCase, but nevermind

**Comment by Lord-Haji** [comment]
bump @HoeenCoder @Mystifi

File: spacialgaze-plugins/misc.js [hunk 9]
```diff
@@ -372,6 +372,15 @@ exports.commands = {
 			msg += '<button class="button" name="send" value="/emote add, ' + target[1] + ', ' + target[2] + '">Add emote</button></center>';
 			delete user.tokens[target[0]];
 			return SG.messageSeniorStaff(msg);
+		case 'disableintroscroll':
+			if (!target[2]) return this.errorReply('/usetoken disableintroscroll, [room]');
+			let roomid = toId(target[1]);
+			if (!Rooms[roomid]) return this.errorReply(`${roomid} is not a room.`);
```
Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js]: Rooms is a function

Comment (Lord-Haji) [reply spacialgaze-plugins/misc.js]: Typo again omg

Comment (HoeenCoder) [reply spacialgaze-plugins/misc.js]: Code like me when i'm tired lol

File: spacialgaze-plugins/misc.js [hunk 10]
```diff
@@ -372,6 +372,15 @@ exports.commands = {
 			msg += '<button class="button" name="send" value="/emote add, ' + target[1] + ', ' + target[2] + '">Add emote</button></center>';
 			delete user.tokens[target[0]];
 			return SG.messageSeniorStaff(msg);
+		case 'disableintroscroll':
+			if (!target[2]) return this.errorReply('/usetoken disableintroscroll, [room]');
```
Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js]: You dont need a target[2] for this command.

File: spacialgaze-plugins/misc.js [hunk 11]
```diff
@@ -398,4 +407,23 @@ exports.commands = {
 		this.parse("/tour create " + target + ", roundrobin");
 	},
 	rtourhelp: ["/rtour [format] - Creates a round robin tournament in the format provided."],
+
+	disableintroscroll: function (target, room, user) {
+		if (!this.can('roomowner')) return false;
+		if (!target) return this.errorReply("No Room Specified");
+		target = toId(target);
+		if (!Rooms(target)) return this.errorReply(`${target} is not a room`);
+		if (Db.disabledScrolls.has(target)) return this.errorReply(`${Rooms(target).title} has roomintro scroll disabled.`);
```
Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js:416]: There should be a check to make sure your not disabling an already disabled intro.

Comment (Lord-Haji) [reply spacialgaze-plugins/misc.js:416]: This line checks that ? isn't it ?

File: spacialgaze-plugins/misc.js [hunk 12]
```diff
@@ -373,7 +373,7 @@ exports.commands = {
 			delete user.tokens[target[0]];
 			return SG.messageSeniorStaff(msg);
 		case 'disableintroscroll':
-			if (!target[2]) return this.errorReply('/usetoken disableintroscroll, [room]');
+			if (!target[1) return this.errorReply('/usetoken disableintroscroll, [room]');
```
Comment (HoeenCoder) [inline spacialgaze-plugins/misc.js]: Your missing a closing `]` now.

File: spacialgaze-plugins/misc.js (hunk 11 above)
Comment (HoeenCoder) [reply spacialgaze-plugins/misc.js:416]: ... 


I feel stupid now.

**Comment by HoeenCoder** [comment]
Need to add /shop side code before merging this

File: rooms.js [hunk 13]
```diff
@@ -1350,7 +1350,8 @@ class ChatRoom extends Room {
 	}
 	getIntroMessage(user) {
 		let message = '';
-		if (this.introMessage) message += '\n|raw|<div class="infobox infobox-roomintro"><div' + (!this.isOfficial ? ' class="infobox-limited"' : '') + '>' + this.introMessage.replace(/\n/g, '') + '</div>';
+		let scroll = Db.DisabledScrolls.get(this.id, false);
```
Comment (Mystifi) [inline rooms.js:1353]: This needs to be changed to lowerCamelCase.

**Pull request merged**